scapy/contrib/isotp/isotp_soft_socket.py

```diff
@@ -0,0 +1,1065 @@
+                if handle is None:
+                    now = TimeoutScheduler._time()
+                    if time_empty is None:
+                        time_empty = now
+                    # 100 ms of grace time before killing the thread
+                    if TimeoutScheduler.GRACE < now - time_empty:
+                        return
+                TimeoutScheduler._wait(handle)
+                TimeoutScheduler._poll()
+
+        finally:
+            # Worst case scenario: if this thread dies, the next scheduled
+            # timeout will start a new one
+            log_runtime.debug("TimeoutScheduler Thread dying @ %f",
+                              TimeoutScheduler._time())
+            TimeoutScheduler._thread = None
+
+    @staticmethod
+    def _poll():
+        # type: () -> None
+        """Execute all the callbacks that were due until now"""
+
+        handles = TimeoutScheduler._handles
+        handle = None
+        while 1:
+            with TimeoutScheduler._mutex:
+                now = TimeoutScheduler._time()
+                if len(handles) == 0 or handles[0]._when > now:
+                    # There is nothing to execute yet
+                    return
+
+                # Time complexity is O(log n)
+                handle = heapq.heappop(handles)
+                callback = None
+                if handle is not None:
+                    callback = handle._cb
+                    handle._cb = True
+
+            # Call the callback here, outside of the mutex
+            if callable(callback):
+                try:
+                    callback()
+                except Exception:
+                    traceback.print_exc()
+
+    @staticmethod
+    def _time():
+        # type: () -> float
+        if six.PY2:
+            return time.time()
+        return time.monotonic()
+
+    class Handle:
+        """Handle for a timeout, consisting of a callback and a time when it
+        should be executed."""
+        __slots__ = ['_when', '_cb']
+
+        def __init__(self,
+                     when,  # type: float
+                     cb  # type: Optional[Union[Callable[[], None], bool]]
+                     ):
+            # type: (...) -> None
+            self._when = when
+            self._cb = cb
+
+        def cancel(self):
+            # type: () -> bool
+            """Cancels this timeout, preventing it from executing its
+            callback"""
+            if self._cb is None:
+                raise Scapy_Exception("cancel() called on "
+                                      "previous canceled Handle")
+            else:
+                if isinstance(self._cb, bool):
+                    # Handle was already executed.
+                    # We don't need to cancel anymore
+                    return False
+                else:
+                    self._cb = None
+                    TimeoutScheduler.cancel(self)
+                    return True
+
+        def __cmp__(self, other):
```

Comment:
Why do you need `__cmp__`?

Reply:
I don't need it, actually. Maybe it was a left over. Removed it. Thanks!